README
README
======

The wavelet sharpen plugin increases apparent sharpness of an image. The amount
of unsharpness in the original image can be taken into account. The plugin is
not compareable to the builtin 'Sharpen' filter of the GIMP, however the
results are simmilar to the ones with the builtin 'Unsharp mask' filter.

LICENCE
-------

Copyright (C) 2008 by Marco Rossini. Distributed under the General Public
License GPLv2+. See the file COPYING which contains the license.

INSTALLATION
------------

See the file INSTALL for instructions how to install the plugins.

USAGE
-----

There are two circumstances that makes it necessary to use the sharpen filter:

1. To reduce the effects due to aliasing of an picture which is in focus.

   a) Set the amount of sharpening using the slider.

   b) You may want to have the checkbox 'Sharpen luminance only' activated in
      order to avoid having colour artifacts (especially in noisy images).

   c) The sharpening radius can be left at 0.5.

2. To sharpen a picture which is out of focus.

   a) Set the amount of sharpening rather high (at first).

   b) Adjust the sharpening radius using the slider to choose the frequency
      part in the image which is to be increased in contrast. This means for
      blurred pictures you may want to set the value high (which sharpens slow-
      ly varying image features, leaving high frequency noise alone). Use low
      values if the image has many details.

   c) Set the sharpening amount back to a value that pleases your eye.

   d) You may want to have the checkbox 'Sharpen luminance only' activated in
      order to avoid having colour artifacts (especially in noisy images).


ChangeLog
ChangeLog
=========

wavelet-sharpen v0.1.2
-------------------------------------------------------------------------------
- Added translations:
  - Polish (thanks to Michal Predotka)
  - Estonian (thanks to Jyri Liira)

wavelet-sharpen v0.1.1
-------------------------------------------------------------------------------
- Added translations:
  - Italian (thanks to Carlo Baldassi)
  - Russian (thanks to Alexandre Prokoudine)

wavelet-sharpen v0.1
-------------------------------------------------------------------------------
- Added gettext support.
- Added language german (by Marco Rossini).
- Source files split.
- Changed build system to use Makefiles.
- Licencse changed from GPLv2 to GPLv2+

wavelet-sharpen v0.1-beta
-------------------------------------------------------------------------------
- First release.


src/colour.c
/* 
 * Wavelet sharpen GIMP plugin
 * 
 * colour.c
 * Copyright 2008 by Marco Rossini
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2+
 * as published by the Free Software Foundation.
 * 
 * Instructions:
 * compile with gimptool, eg. 'gimptool-2.0 --install wavelet-sharpen.c'
 */

#include "plugin.h"

void
rgb2ycbcr (float *r, float *g, float *b)
{
  /* using JPEG conversion here - expecting all channels to be
   * in [0:255] range */
  static float y, cb, cr;
  y = 0.2990 * *r + 0.5870 * *g + 0.1140 * *b;
  cb = -0.1687 * *r - 0.3313 * *g + 0.5000 * *b + 128.0;
  cr = 0.5000 * *r - 0.4187 * *g - 0.0813 * *b + 128.0;
  *r = y;
  *g = cb;
  *b = cr;
}

void
ycbcr2rgb (float *y, float *cb, float *cr)
{
  /* using JPEG conversion here - expecting all channels to be
   * in [0:255] range */
  static float r, g, b;
  r = *y + 1.40200 * (*cr - 128.0);
  g = *y - 0.34414 * (*cb - 128.0) - 0.71414 * (*cr - 128.0);
  b = *y + 1.77200 * (*cb - 128.0);
  *y = r;
  *cb = g;
  *cr = b;
}


src/events.c
/* 
 * Wavelet sharpen GIMP plugin
 * 
 * events.c
 * Copyright 2008 by Marco Rossini
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2+
 * as published by the Free Software Foundation.
 */

#include "plugin.h"
#include "interface.h"

void
reset_channel (GtkWidget * w, gpointer data)
{
  settings.sharpen_amount = 0;
  settings.sharpen_radius = 0.5;
  gtk_adjustment_set_value (GTK_ADJUSTMENT (amount_adj[0]), 0);
  gtk_adjustment_set_value (GTK_ADJUSTMENT (amount_adj[1]), 0.5);
}

void
set_ycbcr_mode (GtkWidget * w, gpointer data)
{
  settings.luminance_only =
    gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (w));
}

void
set_amount (GtkWidget * w, gpointer data)
{
  double val;
  val = gtk_adjustment_get_value (GTK_ADJUSTMENT (w));
  settings.sharpen_amount = val;
}

void
set_radius (GtkWidget * w, gpointer data)
{
  double val;
  val = gtk_adjustment_get_value (GTK_ADJUSTMENT (w));
  settings.sharpen_radius = val;
}


src/plugin.h
/* 
 * Wavelet sharpen GIMP plugin
 * 
 * plugin.h
 * Copyright 2008 by Marco Rossini
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2+
 * as published by the Free Software Foundation.
 */

#ifndef __PLUGIN_H__
#define __PLUGIN_H__

#include <stdlib.h>
#include <math.h>
#include <libgimp/gimp.h>
#include <libgimp/gimpui.h>
#include <libintl.h>
#include <locale.h>

#define LOCALEDIR "/usr/share/locale/"

#define _(string) gettext(string)
#define N_(string) string

#define MAX2(x,y) ((x) > (y) ? (x) : (y))
#define MIN2(x,y) ((x) < (y) ? (x) : (y))
#define CLIP(x,min,max) MAX2((min), MIN2((x), (max)))

#define TIMER_READ (1.0 / 10)
#define TIMER_WRITE (1.0 / 7)
#define TIMER_PROCESS (1.0 - TIMER_READ - TIMER_WRITE)

/* Tool-tips */
#define TT_AMOUNT _("Adjusts the amount of sharpening applied.")
#define TT_RADIUS _("Adjusts the radius of the sharpening. For very unsharp images it is recommended to use higher values. Default is 0.5.")
#define TT_LUMINANCE _("Sharpens luminance only (YCbCr luminance channel). This avoids color artifacts to appear and makes the sharpening faster. Colour sharpness in natural images is not critical for the human eye.")
#define TT_RESET _("Resets to the default values.")

void query (void);
void run (const gchar * name, gint nparams, const GimpParam * param,
		 gint * nreturn_vals, GimpParam ** return_vals);
void wavelet_sharpen (float *fimg[3], unsigned int width,
			     unsigned int height, double amount,
			     double radius, float a, float b);
void sharpen (GimpDrawable * drawable, GimpPreview * preview);
void rgb2ycbcr (float *r, float *g, float *b);
void ycbcr2rgb (float *y, float *cb, float *cr);
void set_ycbcr_mode (GtkWidget * w, gpointer data);
void set_amount (GtkWidget * w, gpointer data);
void set_radius (GtkWidget * w, gpointer data);
gboolean user_interface (GimpDrawable * drawable);
void reset_channel (GtkWidget * w, gpointer data);

extern GimpPlugInInfo PLUG_IN_INFO;

typedef struct
{
  double sharpen_amount;
  double sharpen_radius;
  gboolean luminance_only;
  gboolean show_preview;
  float times[3];
  gint winxsize, winysize;
} wavelet_settings;

extern wavelet_settings settings;

float *fimg[4];
float *buffer[3];
gint channels;

GTimer *timer;

#endif /* __PLUGIN_H__ */


src/sharpen.c
/* 
 * Wavelet sharpen GIMP plugin
 * 
 * sharpen.c
 * Copyright 2008 by Marco Rossini
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2+
 * as published by the Free Software Foundation.
 */

#include "plugin.h"

void
sharpen (GimpDrawable * drawable, GimpPreview * preview)
{
  GimpPixelRgn rgn_in, rgn_out;
  gint i, x1, y1, x2, y2, width, height, x, c;
  guchar *line;
  float val[4], times[3], totaltime;
  int channels_sharpened;

  if (preview)
    {
      gimp_preview_get_position (preview, &x1, &y1);
      gimp_preview_get_size (preview, &width, &height);
      x2 = x1 + width;
      y2 = y1 + height;
    }
  else
    {
      gimp_drawable_mask_bounds (drawable->drawable_id, &x1, &y1, &x2, &y2);
      width = x2 - x1;
      height = y2 - y1;
    }

  gimp_pixel_rgn_init (&rgn_in, drawable, x1, y1, width, height, FALSE,
		       FALSE);
  gimp_pixel_rgn_init (&rgn_out, drawable, x1, y1, width, height,
		       preview == NULL, TRUE);

  /* cache some tiles to make reading/writing faster */
  gimp_tile_cache_ntiles (drawable->width / gimp_tile_width () + 1);

  totaltime = settings.times[0];
  /* FIXME: exclude alpha channel */
  if (settings.luminance_only != TRUE)
    {
      if (channels % 2 == 0)
	totaltime += (channels - 1) * settings.times[1];
      else
	totaltime += channels * settings.times[1];
    }
  else
    totaltime += settings.times[1];
  totaltime += settings.times[2];


  /* FIXME: replace by GIMP functions */
  line = (guchar *) malloc (channels * width * sizeof (guchar));

  /* read the full image from GIMP */
  if (!preview)
    gimp_progress_init (_("Wavelet sharpening..."));
  times[0] = g_timer_elapsed (timer, NULL);
  for (i = 0; i < y2 - y1; i++)
    {
      if (!preview && i % 10 == 0)
	gimp_progress_update (settings.times[0] * i
			      / (double) height / totaltime);
      gimp_pixel_rgn_get_row (&rgn_in, line, x1, i + y1, width);

      /* convert pixel values to float and do colour model conv. */
      for (x = 0; x < width; x++)
	{
	  for (c = 0; c < channels; c++)
	    val[c] = (float) line[x * channels + c];
	  if (channels > 2 && settings.luminance_only == TRUE)
	    rgb2ycbcr (&(val[0]), &(val[1]), &(val[2]));
	  /* save pixel values and scale for sharpening */
	  for (c = 0; c < channels; c++)
	    fimg[c][i * width + x] = val[c] / 255.0;
	  /* FIXME: do gamma correction */
	}
    }
  times[0] = g_timer_elapsed (timer, NULL) - times[0];

  /* sharpen the channels individually */
  times[1] = g_timer_elapsed (timer, NULL);
  /* FIXME: variable abuse (x) */
  channels_sharpened = 0;
  /* FIXME: do not sharpen alpha channel */
  for (c = 0; c < ((channels % 2) ? channels : channels - 1); c++)
    {
      double a, b;
      if (settings.luminance_only == TRUE && c > 0)
	continue;
      buffer[0] = fimg[c];
      b = preview ? 0.0 : settings.times[1] / totaltime;
      a = settings.times[0] + channels_sharpened * settings.times[1];
      a /= totaltime;
      if (settings.sharpen_amount > 0)
	{
	  wavelet_sharpen (buffer, width, height,
			   settings.sharpen_amount,
			   settings.sharpen_radius, a, b);
	  channels_sharpened++;
	}
    }
  times[1] = g_timer_elapsed (timer, NULL) - times[1];
  times[1] /= channels_sharpened;

  /* retransform the image data */
  for (c = 0; c < channels; c++)
    for (i = 0; i < width * height; i++)
      fimg[c][i] = fimg[c][i] * 255;

  /* set alpha to full opacity in single channel preview mode */
  if (preview && channels % 2 == 0)
    for (i = 0; i < width * height; i++)
      fimg[channels - 1][i] = 255;

  /* convert to RGB if necessary */
  if (channels > 2 && settings.luminance_only == TRUE)
    for (i = 0; i < width * height; i++)
      ycbcr2rgb (&(fimg[0][i]), &(fimg[1][i]), &(fimg[2][i]));

  /* clip the values */
  for (c = 0; c < channels; c++)
    {
      for (i = 0; i < width * height; i++)
	{
	  fimg[c][i] = CLIP (fimg[c][i], 0, 255);
	}
    }

  /* write the image back to GIMP */
  times[2] = g_timer_elapsed (timer, NULL);
  for (i = 0; i < height; i++)
    {
      if (!preview)
	gimp_progress_update ((settings.times[0] + channels_sharpened
			       * settings.times[1] + settings.times[2]
			       * i / (double) height) / totaltime);

      /* convert back to guchar */
      for (c = 0; c < channels; c++)
	for (x = 0; x < width; x++)
	  /* avoiding rounding errors !!! */
	  line[x * channels + c] = (guchar) (fimg[c][i * width + x] + 0.5);
      gimp_pixel_rgn_set_row (&rgn_out, line, x1, i + y1, width);
    }
  times[2] = g_timer_elapsed (timer, NULL) - times[2];

  if (!preview)
    {
      settings.times[0] = times[0];
      if (channels_sharpened > 0)
	settings.times[1] = times[1];
      settings.times[2] = times[2];
    }

  /* FIXME: replace by gimp functions */
  free (line);

  if (preview)
    {
      gimp_drawable_preview_draw_region (GIMP_DRAWABLE_PREVIEW (preview),
					 &rgn_out);
      return;
    }
  gimp_drawable_flush (drawable);
  gimp_drawable_merge_shadow (drawable->drawable_id, TRUE);
  gimp_drawable_update (drawable->drawable_id, x1, y1, width, height);
}


src/wavelet.c
/* 
 * Wavelet sharpen GIMP plugin
 * 
 * wavelet.c
 * Copyright 2008 by Marco Rossini
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2+
 * as published by the Free Software Foundation.
 */

#include "plugin.h"

/* code copied from UFRaw (which originates from dcraw) */
void
hat_transform (float *temp, float *base, int st, int size, int sc)
{
  int i;
  for (i = 0; i < sc; i++)
    temp[i] = 2 * base[st * i] + base[st * (sc - i)] + base[st * (i + sc)];
  for (; i + sc < size; i++)
    temp[i] = 2 * base[st * i] + base[st * (i - sc)] + base[st * (i + sc)];
  for (; i < size; i++)
    temp[i] = 2 * base[st * i] + base[st * (i - sc)]
      + base[st * (2 * size - 2 - (i + sc))];
}

/* actual sharpening algorithm */
void
wavelet_sharpen (float *fimg[3], unsigned int width,
		 unsigned int height, double amount, double radius, float a,
		 float b)
{
  float *temp, amt;
  unsigned int i, lev, lpass, hpass, size, col, row;

  size = width * height;

  /* FIXME: replace by GIMP functions */
  temp = (float *) malloc (MAX2 (width, height) * sizeof (float));

  hpass = 0;
  for (lev = 0; lev < 5; lev++)
    {
      if (b != 0)
	gimp_progress_update (a + b * lev / 5.0);
      lpass = ((lev & 1) + 1);
      for (row = 0; row < height; row++)
	{
	  hat_transform (temp, fimg[hpass] + row * width, 1, width, 1 << lev);
	  for (col = 0; col < width; col++)
	    {
	      fimg[lpass][row * width + col] = temp[col] * 0.25;
	    }
	}
      if (b != 0)
	gimp_progress_update (a + b * (lev + 0.4) / 5.0);
      for (col = 0; col < width; col++)
	{
	  hat_transform (temp, fimg[lpass] + col, width, height, 1 << lev);
	  for (row = 0; row < height; row++)
	    {
	      fimg[lpass][row * width + col] = temp[row] * 0.25;
	    }
	}
      if (b != 0)
	gimp_progress_update (a + b * (lev + 0.8) / 5.0);

      amt = amount * exp (-(lev - radius) * (lev - radius) / 1.5) + 1;
      for (i = 0; i < size; i++)
	{
	  fimg[hpass][i] -= fimg[lpass][i];
	  fimg[hpass][i] *= amt;
	
	  if (hpass)
	    fimg[0][i] += fimg[hpass][i];
	}
      hpass = lpass;
    }

  for (i = 0; i < size; i++)
    fimg[0][i] = fimg[0][i] + fimg[lpass][i];

  /* FIXME: replace by GIMP functions */
  free (temp);
}


src/interface.c
/* 
 * Wavelet sharpen GIMP plugin
 * 
 * interface.c
 * Copyright 2008 by Marco Rossini
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2+
 * as published by the Free Software Foundation.
 */

#include "plugin.h"
#include "interface.h"

gboolean
user_interface (GimpDrawable * drawable)
{
  gboolean run;

  gimp_ui_init (_("Wavelet sharpen"), FALSE);

  /* prepare the preview */
  preview = gimp_drawable_preview_new (drawable, &settings.show_preview);
  g_signal_connect_swapped (preview, "invalidated", G_CALLBACK (sharpen),
			    drawable);
  gtk_widget_show (preview);

  /* prepare the amount/radius vbox */
  amount_vbox = gtk_vbox_new (FALSE, 5);
  gtk_widget_show (amount_vbox);

  amount_hbox[0] = gtk_hbox_new (FALSE, 10);
  amount_label[0] = gtk_label_new (_("Amount"));
  gtk_misc_set_alignment (GTK_MISC (amount_label[0]), 0.0, 0.0);
  amount_adj[0] =
    gtk_adjustment_new (settings.sharpen_amount, 0, 10, 0.1, 0.1, 0);
  amount_spin[0] =
    gtk_spin_button_new (GTK_ADJUSTMENT (amount_adj[0]), 0.1, 1);
  amount_scale[0] = gtk_hscale_new (GTK_ADJUSTMENT (amount_adj[0]));
  gtk_scale_set_draw_value (GTK_SCALE (amount_scale[0]), FALSE);
  gtk_box_pack_start (GTK_BOX (amount_vbox), amount_hbox[0], FALSE, FALSE, 0);
  gtk_box_pack_start (GTK_BOX (amount_hbox[0]), amount_label[0], FALSE, FALSE,
		      0);
  gtk_box_pack_start (GTK_BOX (amount_hbox[0]), amount_scale[0], TRUE, TRUE,
		      0);
  gtk_box_pack_start (GTK_BOX (amount_hbox[0]), amount_spin[0], FALSE, FALSE,
		      0);
  g_signal_connect_swapped (amount_adj[0], "value_changed",
			    G_CALLBACK (gimp_preview_invalidate), preview);
  g_signal_connect (amount_adj[0], "value_changed", G_CALLBACK (set_amount),
		    NULL);
  gtk_widget_show (amount_scale[0]);
  gtk_widget_show (amount_spin[0]);
  gtk_widget_show (amount_label[0]);
  gtk_widget_set_tooltip_text (amount_label[0], TT_AMOUNT);
  gtk_widget_set_tooltip_text (amount_scale[0], TT_AMOUNT);
  gtk_widget_set_tooltip_text (amount_spin[0], TT_AMOUNT);
  gtk_widget_show (amount_hbox[0]);

  amount_hbox[1] = gtk_hbox_new (FALSE, 10);
  amount_label[1] = gtk_label_new (_("Radius"));
  gtk_misc_set_alignment (GTK_MISC (amount_label[1]), 0.0, 0.0);
  amount_adj[1] =
    gtk_adjustment_new (settings.sharpen_radius, 0, 2, 0.01, 0.01, 0);
  amount_spin[1] =
    gtk_spin_button_new (GTK_ADJUSTMENT (amount_adj[1]), 0.01, 2);
  amount_scale[1] = gtk_hscale_new (GTK_ADJUSTMENT (amount_adj[1]));
  gtk_scale_set_draw_value (GTK_SCALE (amount_scale[1]), FALSE);
  gtk_box_pack_start (GTK_BOX (amount_vbox), amount_hbox[1], FALSE, FALSE, 0);
  gtk_box_pack_start (GTK_BOX (amount_hbox[1]), amount_label[1], FALSE, FALSE,
		      0);
  gtk_box_pack_start (GTK_BOX (amount_hbox[1]), amount_scale[1], TRUE, TRUE,
		      0);
  gtk_box_pack_start (GTK_BOX (amount_hbox[1]), amount_spin[1], FALSE, FALSE,
		      0);
  g_signal_connect_swapped (amount_adj[1], "value_changed",
			    G_CALLBACK (gimp_preview_invalidate), preview);
  g_signal_connect (amount_adj[1], "value_changed", G_CALLBACK (set_radius),
		    NULL);
  gtk_widget_show (amount_scale[1]);
  gtk_widget_show (amount_spin[1]);
  gtk_widget_show (amount_label[1]);
  gtk_widget_set_tooltip_text (amount_label[1], TT_RADIUS);
  gtk_widget_set_tooltip_text (amount_scale[1], TT_RADIUS);
  gtk_widget_set_tooltip_text (amount_spin[1], TT_RADIUS);
  gtk_widget_show (amount_hbox[1]);

  /* prepare luminance select */
  if (channels > 2)
    {
      lum_toggle = gtk_check_button_new_with_label (_("Sharpen luminance only"));
      if (settings.luminance_only == TRUE)
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (lum_toggle), 1);
      else
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (lum_toggle), 0);
      g_signal_connect (lum_toggle, "toggled",
			G_CALLBACK (set_ycbcr_mode), NULL);
      g_signal_connect_swapped (lum_toggle, "toggled",
				G_CALLBACK (gimp_preview_invalidate),
				preview);
      gtk_widget_set_tooltip_text (lum_toggle, TT_LUMINANCE);
      gtk_widget_show (lum_toggle);
    }
  else
    {
      lum_toggle = NULL;
    }

  /* prepare the reset buttons */
  button_hbox = gtk_hbox_new (FALSE, 10);
  reset_button = gtk_button_new_from_stock (GIMP_STOCK_RESET);
  g_signal_connect (reset_button, "clicked", G_CALLBACK (reset_channel),
		    NULL);
  if (lum_toggle)
    gtk_box_pack_start (GTK_BOX (button_hbox), lum_toggle, TRUE, TRUE, 0);
  gtk_widget_set_tooltip_text (reset_button, TT_RESET);
  gtk_box_pack_start (GTK_BOX (button_hbox), reset_button, FALSE, FALSE, 0);
  gtk_widget_show (reset_button);

  gtk_widget_show (button_hbox);

  /* prepeare the dialog */
  dialog_vbox = gtk_vbox_new (FALSE, 10);
  gtk_container_set_border_width (GTK_CONTAINER (dialog_vbox), 10);
  dialog = gimp_dialog_new (_("Wavelet sharpen"), "wavelet sharpen", NULL, 0,
			    gimp_standard_help_func,
			    "plug-in-wavelet-sharpen", GTK_STOCK_CANCEL,
			    GTK_RESPONSE_CANCEL, GTK_STOCK_OK,
			    GTK_RESPONSE_OK, NULL);
  gtk_container_add (GTK_CONTAINER (GTK_DIALOG (dialog)->vbox), dialog_vbox);
  gtk_box_pack_start (GTK_BOX (dialog_vbox), preview, TRUE, TRUE, 0);
  gtk_box_pack_start (GTK_BOX (dialog_vbox), amount_vbox, FALSE, FALSE, 0);
  gtk_box_pack_start (GTK_BOX (dialog_vbox), button_hbox, FALSE, FALSE, 0);

  gtk_window_set_position (GTK_WINDOW (dialog), GTK_WIN_POS_CENTER);
  if (settings.winxsize > 0 && settings.winysize > 0)
    gtk_window_resize (GTK_WINDOW (dialog), settings.winxsize,
		       settings.winysize);

  gtk_widget_show (dialog_vbox);
  gtk_widget_show (dialog);

  run = (gimp_dialog_run (GIMP_DIALOG (dialog)) == GTK_RESPONSE_OK);

  gtk_window_get_size (GTK_WINDOW (dialog), &(settings.winxsize),
		       &(settings.winysize));

  /* FIXME: destroy all widgets - memory leak! */
  gtk_widget_destroy (dialog);

  return run;
}


src/interface.h
/* 
 * Wavelet sharpen GIMP plugin
 * 
 * wavelet-sharpen.c
 * Copyright 2008 by Marco Rossini
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2+
 * as published by the Free Software Foundation.
 * 
 * Instructions:
 * compile with gimptool, eg. 'gimptool-2.0 --install wavelet-sharpen.c'
 */

#include "plugin.h"

/* declare all widgets global to make things a lot easier */

/* colour mode frame */
GtkWidget *lum_toggle;

/* amount/radius frame */
GtkWidget *amount_label[2], *amount_spin[2];
GtkWidget *amount_hbox[2], *amount_vbox, *amount_scale[2];
GtkObject *amount_adj[2];

/* reset buttons */
GtkWidget *reset_button, *button_hbox;

/* dialog */
GtkWidget *dialog, *dialog_vbox;
GtkWidget *preview;


src/plugin.c
/* 
 * Wavelet sharpen GIMP plugin
 * 
 * plugin.c
 * Copyright 2008 by Marco Rossini
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2+
 * as published by the Free Software Foundation.
 */

#include "plugin.h"

GimpPlugInInfo PLUG_IN_INFO = { NULL, NULL, query, run };

MAIN ()
     void query (void)
{
  static GimpParamDef args[] = {
    {GIMP_PDB_INT32, "run-mode", "Run mode"},
    {GIMP_PDB_IMAGE, "image", "Input image"},
    {GIMP_PDB_DRAWABLE, "drawable", "Input drawable"},
    {GIMP_PDB_FLOAT, "amount", "Amount of sharpening"},
    {GIMP_PDB_FLOAT, "radius", "Radius of sharpening"},
    {GIMP_PDB_INT8, "luminance", "Sharpen luminance channel only"}
  };

  gimp_install_procedure ("plug-in-wavelet-sharpen",
			  _("Sharpens the image using wavelets."),
			  _("Sharpens the image using wavelets."),
			  /* Translators: this is [firstname surname] */
			  _("Marco Rossini"),
			  _("Copyright 2008 Marco Rossini"),
			  "2008",
			  _("_Wavelet sharpen ..."),
			  "RGB*, GRAY*",
			  GIMP_PLUGIN, G_N_ELEMENTS (args), 0, args, NULL);

  gimp_plugin_domain_register("gimp20-wavelet-sharpen-plug-in", LOCALEDIR);

  gimp_plugin_menu_register ("plug-in-wavelet-sharpen",
			     "<Image>/Filters/Enhance");
}

wavelet_settings settings = {
  0,				/* sharpen_amount */
  0.5,				/* sharpen_radius */
  TRUE,				/* luminance_only */
  TRUE,				/* show_preview */
  {1.46, 2.94, 1.5},		/* times */
  -1, -1			/* winxsize, winysize */
};

void
run (const gchar * name, gint nparams, const GimpParam * param,
     gint * nreturn_vals, GimpParam ** return_vals)
{
  static GimpParam values[1];
  GimpRunMode run_mode;
  GimpDrawable *drawable;
  gint i;

  bindtextdomain("gimp20-wavelet-sharpen-plug-in", LOCALEDIR);
  textdomain("gimp20-wavelet-sharpen-plug-in");
  bind_textdomain_codeset("gimp20-wavelet-sharpen-plug-in", "UTF-8");

  timer = g_timer_new ();

  /* Setting mandatory output values */
  *nreturn_vals = 1;
  *return_vals = values;
  values[0].type = GIMP_PDB_STATUS;
  values[0].data.d_status = GIMP_PDB_SUCCESS;

  /* restore settings saved in GIMP core */
  gimp_get_data ("plug-in-wavelet-sharpen", &settings);

  drawable = gimp_drawable_get (param[2].data.d_drawable);
  channels = gimp_drawable_bpp (drawable->drawable_id);

  /* allocate buffers */
  /* FIXME: replace by GIMP funcitons */
  for (i = 0; i < channels; i++)
    {
      fimg[i] = (float *) malloc (drawable->width * drawable->height
				  * sizeof (float));
    }
  buffer[1] = (float *) malloc (drawable->width * drawable->height
				* sizeof (float));
  buffer[2] = (float *) malloc (drawable->width * drawable->height
				* sizeof (float));

  /* run GUI if in interactiv mode */
  run_mode = param[0].data.d_int32;
  if (run_mode == GIMP_RUN_INTERACTIVE)
    {
      if (!user_interface (drawable))
	{
	  gimp_drawable_detach (drawable);
	  /* FIXME: should return error status here */
	  return;
	}
    }
  else
    {
      if (nparams == 6)
	{
	  settings.sharpen_amount = param[3].data.d_float;
	  settings.sharpen_radius = param[4].data.d_float;
	  settings.luminance_only = !!(param[5].data.d_int8);
	}
    }

  sharpen (drawable, NULL);

  /* free buffers */
  /* FIXME: replace by GIMP functions */
  for (i = 0; i < channels; i++)
    {
      free (fimg[i]);
    }
  free (buffer[1]);
  free (buffer[2]);

  gimp_displays_flush ();
  gimp_drawable_detach (drawable);

  /* save settings in the GIMP core */
  gimp_set_data ("plug-in-wavelet-sharpen", &settings,
		 sizeof (wavelet_settings));
}
